File: python/construct-binary-tree-from-preorder-and-inorder-traversal.py
# 105. 从前序与中序遍历序列构造二叉树 - Medium

from typing import List


class TreeNode:
    def __init__(self, val, left = None, right = None) -> None:
        self.val = val
        self.left = left
        self.right = right

class Solution:

    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def buildTreeHelper(pLeft: int, pRight: int, iLeft: int, iRight: int) -> TreeNode:
            if pLeft > pRight:
                return None
            root = TreeNode(preorder[pLeft])
            rootIdx = index[preorder[pLeft]]
            leftSize = rootIdx - iLeft
            root.left = buildTreeHelper(pLeft + 1, pLeft + leftSize, iLeft, rootIdx - 1)
            root.right = buildTreeHelper(pLeft + leftSize + 1, pRight, rootIdx + 1, iRight)
            return root
             
        index = { val: i for i, val in enumerate(inorder)}
        return buildTreeHelper(0, len(preorder) - 1, 0, len(inorder) - 1)

File: go/construct-binary-tree-from-preorder-and-inorder-traversal.go
// 105. 从前序与中序遍历序列构造二叉树 - Medium
// https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
package leetcode

func buildTree(preorder []int, inorder []int) *TreeNode {
	index := make(map[int]int)
	for i, v := range inorder {
		index[v] = i
	}
	var buildHelper func(pLeft int, pRight int, iLeft int, iRight int) *TreeNode
	buildHelper = func(pLeft int, pRight int, iLeft int, iRight int) *TreeNode {
		if pLeft > pRight {
			return nil
		}
		root := &TreeNode{Val: preorder[pLeft], Left: nil, Right: nil}
		rootIdx := index[preorder[pLeft]]
		var leftSize int = rootIdx - iLeft
		root.Left := buildHelper(pLeft + 1, pLeft + leftSize, iLeft, rootIdx - 1)
		root.Right := buildHelper(pLeft + leftSize + 1, pRight, rootIdx + 1, iRight)
		return root
	}
	
	return buildHelper(0, len(preorder)-1, 0, len(inorder)-1)
}
